pkg/controller/machine/machine_controller.go

```diff
@@ -69,6 +59,12 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// Store the controller on the ReconcileMachine instance
+	// to dynamically add watchers to external objects at runtime.
+	if r, ok := r.(*ReconcileMachine); ok {
```

Comment:
What if we made this whole init process cleaner?

```go
func Add(mgr manager.Manager) error {
	r := &ReconcileMachine{
		Client:   mgr.GetClient(),
		scheme:   mgr.GetScheme(),
	}

	c, err := controller.New("machine_controller", mgr, controller.Options{Reconciler: r})
	if err != nil {
		return err
	}
	
	r.controller = c

	// Watch for changes to Machine
	return c.Watch(
		&source.Kind{Type: &clusterv1.Machine{}},
		&handler.EnqueueRequestForObject{},
	)
}
```

Reply:
I tried this approach yesterday and went back to the standard one, mostly because it doesn't break tests and it's familiar for folks already working with kubebuilder.

Reply:
I strongly prefer what I have above. It's significantly easier to understand and removes the indirection we have now. Anyone else have opinions? @detiber @asauber @davidewatson @dlipovetsky

Reply:
Can we open an issue instead and discuss pros/cons there? It also requires to rewrite the testing suite, which can be part of a follow-up PR if we feel strongly about it